gc/base/MemoryPoolAddressOrderedList.hpp

```diff
@@ -146,13 +162,104 @@ class MM_MemoryPoolAddressOrderedList : public MM_MemoryPoolAddressOrderedListBa
+	/**
+	 * @return the recorded estimate of scannable in the receiver
+	 */
+	MMINLINE uintptr_t getScannableBytes() { return _scannableBytes; }
+	/**
+	 * @return the recorded estimate of non-scannable in the receiver
+	 */
+	MMINLINE uintptr_t getNonScannableBytes() { return _nonScannableBytes; }
+
+	/**
+	 * remove a free entry from freelist
+	 */
+	void removeFromFreeList(void *addrBase, void *addrTop, MM_HeapLinkedFreeHeader *previousFreeEntry, MM_HeapLinkedFreeHeader *nextFreeEntry)
+	{
+		bool const compressed = compressObjectReferences();
+		uintptr_t freeEntrySize = ((uintptr_t)addrTop) - ((uintptr_t)addrBase);
+		MM_HeapLinkedFreeHeader::fillWithHoles(addrBase, freeEntrySize, compressed);
+		if (previousFreeEntry) {
+			previousFreeEntry->setNext(nextFreeEntry, compressed);
+		}else {
+			_heapFreeList = nextFreeEntry;
+		}
+	}
+
+	void fillWithHoles(void *addrBase, void *addrTop)
+	{
+		bool const compressed = compressObjectReferences();
+		MM_HeapLinkedFreeHeader::fillWithHoles(addrBase, ((uintptr_t)addrTop) - ((uintptr_t)addrBase), compressed);
+	}
+
+	MMINLINE uintptr_t getABSDarkMatterBytes()
+	{
+		return _darkMatterBytes;
+	}
+
+	MMINLINE uintptr_t getABSActualFreeMemorySize()
+	{
+		return _freeMemorySize;
+	}
+
+	MMINLINE uintptr_t getAdjustedbytesForCardAlignment()
+	{
+		return _adjustedbytesForCardAlignment;
+	}
+
+	MMINLINE void setAdjustedbytesForCardAlignment(uintptr_t adjustedbytes, bool needSync)
+	{
+		if (needSync) {
+			MM_AtomicOperations::add(&(_adjustedbytesForCardAlignment), adjustedbytes);
+		} else {
+			_adjustedbytesForCardAlignment += adjustedbytes;
+		}
+	}
+
+	MMINLINE void initialFirstUnAlignmentFreeEntry()
+	{
+		_firstUnAlignmentFreeEntry = (NULL == _heapFreeList) ? (MM_HeapLinkedFreeHeader *) UDATA_MAX : _heapFreeList;
+		_preFirstUnAlignmentFreeEntry =  (MM_HeapLinkedFreeHeader *) UDATA_MAX;
+	}
+
+	MMINLINE void resetFirstUnAlignmentFreeEntry()
+	{
+		_firstUnAlignmentFreeEntry =  (MM_HeapLinkedFreeHeader *) UDATA_MAX;
+		_preFirstUnAlignmentFreeEntry =  (MM_HeapLinkedFreeHeader *) UDATA_MAX;
+	}
+
+	MMINLINE virtual uintptr_t getDarkMatterBytes()
+	{
+		return _darkMatterBytes + _adjustedbytesForCardAlignment;
+	}
+
+	MMINLINE virtual uintptr_t getActualFreeMemorySize()
+	{
+		return _freeMemorySize - _adjustedbytesForCardAlignment;
+	}
+
+	MMINLINE uintptr_t getFreeMemoryAndDarkMatterBytes() {
+		return getActualFreeMemorySize() + getDarkMatterBytes();
+	}
+
 	/**
 	 * Create a MemoryPoolAddressOrderedList object.
 	 */
 	MM_MemoryPoolAddressOrderedList(MM_EnvironmentBase *env, uintptr_t minimumFreeEntrySize) :
 		MM_MemoryPoolAddressOrderedListBase(env, minimumFreeEntrySize)
 		,_heapFreeList(NULL)
 		,_largeObjectCollectorAllocateStats(NULL)
+		,_scannableBytes(0)
+		,_nonScannableBytes(0)
+		,_firstUnAlignmentFreeEntry((MM_HeapLinkedFreeHeader *) UDATA_MAX)
+		,_preFirstUnAlignmentFreeEntry((MM_HeapLinkedFreeHeader *) UDATA_MAX)
```

Comment:
(MM_HeapLinkedFreeHeader *) UDATA_MAX is frequently used. For easier code reading you can define a const FREE_ENTRY_END or FREE_ENTRY_MAX:

#define FREE_ENTRY_END ((MM_HeapLinkedFreeHeader *) UDATA_MAX)

Reply:
changed